woocommerce/woocommerce-ios
# [Order Editing][1/3] Prepare view model for optimistic updates for shipping and billing addresses

As part of **#5339**

**Part 1 of #5610**

Other parts:
- [**[2/3] Enable optimistic updates for shipping address in the Order details screen.**](https://github.com/woocommerce/woocommerce-ios/pull/6727)
- [**[3/3] Enable optimistic updates for billing address in the Order details screen.**](https://github.com/woocommerce/woocommerce-ios/pull/6728)

### Description

This is the first of three tasks. This PR aims to prepare the view model for optimistic updates. It must have some considerations to be ready, like the feature flag to enable/disable the optimistic updates, manage the notices, and execute the update action. I've followed the same approach for [updating the customer note optimistically](https://github.com/woocommerce/woocommerce-ios/pull/6697). 

In this case, the view model is `EditOrderAddressFormViewModel`. This view model conforms to the protocol `AddressFormViewModelProtocol` and is inherited from the class `AddressFormViewModel`. The parent class holds shared sync/management logic. 

I included the logic to manage the feature flag in the parent class. So, I've added a `FeatureFlagService` as a dependency of this class and created a computed property to indicate whether the optimistic updates are available or not.

Also, there was a mix of responsibilities of who managed the notices; some notices were managed by the parent class and others by the view model. I decided to move to the parent class to simplify the logic. 

Like we did for the customer notes, the view model handles the update action and decides which strategy should use based on the feature flag value. And I've added some methods needed to update the `order` of this view model and to synchronise the form fields when the order is updated.

In addition, I've created a new feature flag (`useUpdateOrderAddressOptimisticallyIfAvailable`), only available for development. The idea of this flag is to be sure that this change doesn't affect other devs and the production code because the UI is not ready for optimistic updates yet. This feature flag will be removed when development to enable optimistic updating of the address of an order is finished. 

### Testing instructions

I've updated the unit tests related to this view model and added new ones. However, the UI is not ready for optimistic updates yet, so I think we should check it in the next tasks. 

---
